File: webpackage/admin/opengraph.js
function extractOpenGraphFromDocument(doc) {
  const els = doc.head.querySelectorAll("meta[property^='og:']");
  const tags = {};

  for (let tag of els)
    tags[tag.attributes.property.value.substr(3)] = tag.content;
  return tags;
}

export function fetchOpenGraphFromUrl(url) {
  return fetch(`/admin/opengraph?url=${encodeURIComponent(url)}`).then(response => {
    return response.text();
  }).then(html => {
    const parser = new DOMParser();
    const data = extractOpenGraphFromDocument(
      new DOMParser().parseFromString(html, 'text/html')
    );

    if (!data.url) data.url = url;
    return data;
  });
}

export function createOpenGraphView(data) {
  console.log("createOpenGraphView from", data);
  if (window.cms && window.cms.createOpenGraphView)
    return window.cms.openGraphView(block);
  else {
    const wrapper = document.createElement("figure");
    const root = document.createElement("a");

    root.className = "opengraph-view";
    root.href = data.url;
    root.target = "_blank";
    if (data.type)
      root.dataset.type = data.type;
    if (data.image) {
      const image = document.createElement("img");
      image.src = data.image;
      image.className = "opengraph-thumbnail";
      if (data["image:width"]) { image.width = data["image:width"]; }
      if (data["image:height"]) { image.height = data["image:height"]; }
      root.appendChild(image);
    }
    if (data.title) {
      const title = document.createElement("h4");
      title.textContent = data.title;
      title.className = "opengraph-title";
      root.appendChild(title);
    }
    if (data.description) {
      const description = document.createElement("div");
      description.textContent = data.description;
      description.className = "opengraph-description";
      root.appendChild(description);
    }
    wrapper.appendChild(root);
    console.log("OpenGraphView", root);
    return wrapper.innerHTML;
  }
}

File: libcrails-cms/crails/cms/opengraph.cpp
#include "opengraph.hpp"
#include <crails/html_template.hpp>

using namespace Crails;
using namespace Crails::Cms;
using namespace std;

static string type_to_string(OpenGraph::Type type)
{
  switch (type)
  {
  case OpenGraph::ArticleType:
    return "article";
  default:
    break ;
  }
  return "website";
}

string OpenGraph::render() const
{
  stringstream stream;

  stream
    << HtmlTemplate::tag("meta", map<string,string>{
      {"property", "og:title"}, {"content", title}
    })
    << HtmlTemplate::tag("meta", map<string,string>{
      {"property", "og:type"}, {"content", type_to_string(type)}
    })
    << HtmlTemplate::tag("meta", map<string,string>{
      {"property", "og:url"}, {"content", url}
    })
    << HtmlTemplate::tag("meta", map<string,string>{
      {"property", "og:description"}, {"content", description}
    });
  if (locales.size() > 0)
  {
    stream << HtmlTemplate::tag("meta", map<string,string>{
      {"property", "og:locale"}, {"content", *locales.begin()}
    });
    for (auto locale : locales)
    {
      stream << HtmlTemplate::tag("meta", map<string,string>{
        {"property", "og:locale:alternate"}, {"content", locale}
      });
    }
  }
  if (image.length() > 0)
  {
    stream << HtmlTemplate::tag("meta", map<string,string>{
      {"property", "og:image"}, {"content", image}
    });
  }
  return stream.str();
}
